### PR TITLE
feat: BitPackedCompressor allows signed arrays

### PR DESCRIPTION
**Pull request by a10y**
Most of the work to support signed integers has been done in BitPackedArray.

This PR removes some assertions and branches in the compressor to make it possible to bit-pack an array of signed ints.

### CODE REVIEW

**Comment by gatesn** [comment]
Is it worth lumping in the change to disallow a no-op FoR array too?

Or better yet, disallow FoR when it has minimal impact on bitpacking width.

**Comment by a10y** [comment]
I've changed the condition in `FoRCompressor::can_compress` to now always reject if min=0 and shift=0. Previously if the type was signed it would not bail.

I did not change FoRCompressor to determine how many bits of savings is enough to make up for the extra unfor step. I don't have a good intuition on what that would be.

File: docs/quickstart.rst [hunk 1]
```diff
@@ -46,9 +46,9 @@ Use :func:`~vortex.encoding.compress` to compress the Vortex array and check the
 
    >>> cvtx = vortex.compress(vtx)
    >>> cvtx.nbytes
-   16756
+   15243
```
Comment (a10y) [inline docs/quickstart.rst]: hey look numbers go down

**Comment by lwwmanning** [comment]
> I've changed the condition in `FoRCompressor::can_compress` to now always reject if min=0 and shift=0. Previously if the type was signed it would not bail.
> 
> I did not change FoRCompressor to determine how many bits of savings is enough to make up for the extra unfor step. I don't have a good intuition on what that would be.

I can live with excluding if min and shift are zero (as a way to short-circuit), but it's a bit unnecessary. The sampling compressor already penalizes extra array depth. So FoR(0,0) + Bitpacking on signed ints will be strictly worse than just Bitpacking and won't be chosen anyway

File: encodings/fastlanes/src/bitpacking/mod.rs [hunk 2]
```diff
@@ -64,10 +64,6 @@ impl BitPackedArray {
     ) -> VortexResult<Self> {
         let dtype = DType::Primitive(ptype, validity.nullability());
 
-        if !dtype.is_unsigned_int() {
```
Comment (lwwmanning) [inline encodings/fastlanes/src/bitpacking/mod.rs:67]: should check `is_int` instead of removing check

**Comment by gatesn** [comment]
> > I've changed the condition in `FoRCompressor::can_compress` to now always reject if min=0 and shift=0. Previously if the type was signed it would not bail.
> > I did not change FoRCompressor to determine how many bits of savings is enough to make up for the extra unfor step. I don't have a good intuition on what that would be.
> 
> I can live with excluding if min and shift are zero (as a way to short-circuit), but it's a bit unnecessary. The sampling compressor already penalizes extra array depth. So FoR(0,0) + Bitpacking on signed ints will be strictly worse than just Bitpacking and won't be chosen anyway

By this logic we shouldn't exclude any compressors based on metrics, e.g. run-end encoding should ignore the avg_run_length.

**Comment by lwwmanning** [comment]
> > > I've changed the condition in `FoRCompressor::can_compress` to now always reject if min=0 and shift=0. Previously if the type was signed it would not bail.
> > > I did not change FoRCompressor to determine how many bits of savings is enough to make up for the extra unfor step. I don't have a good intuition on what that would be.
> > 
> > 
> > I can live with excluding if min and shift are zero (as a way to short-circuit), but it's a bit unnecessary. The sampling compressor already penalizes extra array depth. So FoR(0,0) + Bitpacking on signed ints will be strictly worse than just Bitpacking and won't be chosen anyway
> 
> By this logic we shouldn't exclude any compressors based on metrics, e.g. run-end encoding should ignore the avg_run_length.

Yeah, I think that's correct. More importantly, we should also filter codecs based on whether `can_compress` the full array, THEN do the sampling compression, then compress-like the best on sample. Right now, we check `can_compress` on the sample, which may fail on the full array, causing a resample... which could fail arbitrarily many times. 

E.g., if we theoretically implemented BitPackedArray::can_compress for signed ints by checking if `min >= 0` on an array with 1 million integers, of which only 1 was negative, we'd expect it to have to resample ~1000 times on average in order to successfully compress.

**Review by lwwmanning — APPROVED** [review]

**Pull request merged**